src/Servers/Kestrel/Core/src/Internal/TlsConnectionFeature.cs

```diff
@@ -99,7 +103,27 @@ public int KeyExchangeStrength
 
         public Task<X509Certificate2?> GetClientCertificateAsync(CancellationToken cancellationToken)
         {
-            return Task.FromResult(ClientCertificate);
+            // Only try once per connection
+            if (_clientCertTask != null)
+            {
+                return _clientCertTask;
+            }
+
+            if (ClientCertificate != null
+                || ClientCertificateMode != ClientCertificateMode.DelayCertificate
+                // Delayed client cert negotiation is not allowed on HTTP/2 (or HTTP/3, but that's implemented elsewhere).
+                || _sslStream.NegotiatedApplicationProtocol == SslApplicationProtocol.Http2)
```

Comment:
Do we have a test to verify we aren't renegotiating when NegotiatedApplicationProtocol is HTTP/2?

Reply:
Added in the HTTP/2 HttpClient interop tests. That's one of the only places we test HTTP/2 over TLS.